### DISCUSSION TITLE
Lesson%202%3A%20Lets%20Go%203D%21/1.%203D%20Cube/

### DISCUSSION THREAD

**Discussion by giscus** [wmarti/MetalTutorial#6] [Announcements]
# Lesson%202%3A%20Lets%20Go%203D%21/1.%203D%20Cube/

https://metaltutorial.com/Lesson%202%3A%20Lets%20Go%203D%21/1.%203D%20Cube/

**Comment by CaptainXX** [comment]
I got these errors during compilation:

```sh
In file included from /Users/cooper/Projects/pcl_analyzer/src/mtl_engine.mm:1:
In file included from /Users/cooper/Projects/pcl_analyzer/src/mtl_engine.hpp:6:
/Users/cooper/Projects/pcl_analyzer/3rd_party/metal-cpp-cmake/metal-cmake/metal-cpp/Metal/MTLRenderPass.hpp:250:90: error: 
      no type named 'SamplePosition' in namespace 'MTL'
    void                                                   setSamplePositions(const MTL::SamplePosition* positions, NS::UInteger count);
                                                                                    ~~~~~^
/Users/cooper/Projects/pcl_analyzer/3rd_party/metal-cpp-cmake/metal-cmake/metal-cpp/Metal/MTLRenderPass.hpp:252:84: error: 
      no type named 'SamplePosition' in namespace 'MTL'
    NS::UInteger                                           getSamplePositions(MTL::SamplePosition* positions, NS::UInteger count);
                                                                              ~~~~~^
/Users/cooper/Projects/pcl_analyzer/3rd_party/metal-cpp-cmake/metal-cmake/metal-cpp/Metal/MTLRenderPass.hpp:760:75: error: 
      no type named 'SamplePosition' in namespace 'MTL'
_MTL_INLINE void MTL::RenderPassDescriptor::setSamplePositions(const MTL::SamplePosition* positions, NS::UInteger count)
                                                                     ~~~~~^
/Users/cooper/Projects/pcl_analyzer/3rd_party/metal-cpp-cmake/metal-cmake/metal-cpp/Metal/MTLRenderPass.hpp:766:77: error: 
      no type named 'SamplePosition' in namespace 'MTL'
_MTL_INLINE NS::UInteger MTL::RenderPassDescriptor::getSamplePositions(MTL::SamplePosition* positions, NS::UInteger count)
                                                                       ~~~~~^
4 errors generated.
```

And this can be solved by adding this line at the top of mtl_engine.hpp.

```c++
#include "Metal/MTLTypes.hpp"
```

Hope this helps if anyone encountered this error. :D

**Reply by gjhfhj** [reply]
Hi CREATER,

I noticed a small issue with the view matrix formula in the tutorial. The matrix you provided doesn't seem to match the standard conventions for constructing a view matrix from the camera's forward, right, up, and position vectors.

For example, based on the vector  F=(1,0,0), R=(0,1,0) U=(0,0,1) and P = (0,1,1), the correct view matrix should look like:
                 0    1     0     -1
Mview =   0    0     -1     1
                -1    0     0      0
                 0    0     0      1
It seems there might be a mix-up in the signs or the order of the vectors in your matrix. I thought it might be helpful to point this out, as this will affect the way the scene is rendered.

Thanks for the great tutorial! Looking forward to more content from you.

Best

**Comment by MERL10N** [comment]
Amazing tutorial so far! It has been a great resource for me to create a game engine using metal-cpp. How is the progress on the lighting and 3D Models tutorial?

**Reply by wmarti** [reply]
Thanks for the support! Much appreciated :) Unfortunately a lot of my time has been eaten up by work the past few months. It may take some time for me to get those chapters right, but I will get there so stay tuned.

**Reply by MERL10N** [reply]
Hi,
No worries and no rush. I rather you take all the time needed to make a quality tutorial for the later chapters. I will be looking forward to when they come out. Cheers!  :)

**Comment by sftrabbit** [comment]
Do we need any synchronisation around that memcpy that's copying the transformation data to the GPU? Is it not in theory possible that the GPU might be reading from that buffer at the same time that we're copying into it for the next frame? I've seen other examples deal with this by multiplying the size of buffers by N, where N is the number of frames in-flight, and then writing/reading to sections of the buffers based on which frame is currently being rendered.

**Comment by Plegeus** [comment]
I am not sure the data in *Step 3: View Transformation* is correct.
`M_view[0][0] == 1` supposedly, but that cannot be if `R == [0, 1, 0]`, given `R_x == R[0]`.

Perhaps the order was meant to be R, U, F?

I am not sure, this is all new to me.

**Reply by Plegeus** [reply]
Otherwise an awesome tutorial!
